### PR TITLE
Remove global use for magic

### PR DESCRIPTION
**Pull request by iotaaxel**
- [x] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

- The global variable was marked and moved. Will now use local static variable instead.

### CODE REVIEW

**Comment by trufae** [comment]
local static is the same as global with a smaller scope. that doesn't solves the problem, the variables marked as R_TH_LOCAL are all global whatever context they are defined

File: libr/magic/apprentice.c [hunk 1]
```diff
@@ -169,10 +169,14 @@ static int get_type(const char *l, const char **t) {
 	return p->type;
 }
 
-static void init_file_tables(void) {
-	static R_TH_LOCAL bool done = false;
+static void init_file_tables(bool *done) {
+	// Initialize
+	if (done == NULL) {
```
Comment (trufae) [inline libr/magic/apprentice.c:174]:
```suggestion
	if (!done) {
```

File: libr/magic/apprentice.c [hunk 2]
```diff
@@ -169,10 +169,14 @@ static int get_type(const char *l, const char **t) {
...
+		done = false;
+	}
+
 	const struct type_tbl_s *p;
-	if (done) {
+	if (&done) {
```
Comment (trufae) [inline libr/magic/apprentice.c:179]:
```suggestion
	if (*done) {
```

File: libr/magic/apprentice.c [hunk 3]
```diff
@@ -276,8 +280,9 @@ struct mlist * file_apprentice(RMagic *ms, const char *fn, size_t fn_size, int a
 	char *p, *mfn;
 	int file_err, errs = -1;
 	struct mlist *mlist;
+	R_TH_LOCAL bool *done = NULL;
```
Comment (trufae) [inline libr/magic/apprentice.c:283]:
```suggestion
	bool done = false;
```

File: libr/magic/apprentice.c [hunk 4]
```diff
@@ -276,8 +280,9 @@ struct mlist * file_apprentice(RMagic *ms, const char *fn, size_t fn_size, int a
...
 
-	init_file_tables ();
+	init_file_tables (done);
```
Comment (trufae) [inline libr/magic/apprentice.c:285]:
```suggestion
	init_file_tables (&done);
```

File: libr/magic/mdump.c [hunk 5]
```diff
@@ -45,7 +45,7 @@
 
 #ifndef COMPILE_ONLY
 void file_mdump(struct r_magic *m) {
-	static R_TH_LOCAL const char optyp[] = { FILE_OPS };
+	static const char optyp[] = { FILE_OPS };
```
Comment (trufae) [inline libr/magic/mdump.c:48]: keep the R_TH_LOCAL. thats a global
```suggestion
	static R_TH_LOCAL const char optyp[] = { FILE_OPS };
```

File: libr/magic/mdump.c [hunk 6]
```diff
@@ -196,9 +196,9 @@ const char *file_fmttime(ut32 v, int local, char *pp) {
 		r_ctime_r (&t, pp);
 	} else {
 #ifndef HAVE_DAYLIGHT
-		static R_TH_LOCAL int daylight = 0;
+		static int daylight = 0;
```
Comment (trufae) [inline libr/magic/mdump.c:199]:
```suggestion
		static R_TH_LOCAL int daylight = 0;
```

File: libr/magic/mdump.c [hunk 7]
```diff
@@ -196,9 +196,9 @@ const char *file_fmttime(ut32 v, int local, char *pp) {
...
 #ifdef HAVE_TM_ISDST
-		static R_TH_LOCAL time_t now = (time_t)0;
+		static time_t now = (time_t)0;
```
Comment (trufae) [inline libr/magic/mdump.c:201]:
```suggestion
		static R_TH_LOCAL time_t now = (time_t)0;
```

**Review by trufae — CHANGES_REQUESTED** [review]
Remember not to remove the R_TH_LOCAL unless you can also remove the `static` so the global is gone

**Comment by iotaaxel** [comment]
I am getting some CI failure messages.

File: libr/magic/apprentice.c [hunk 8]
```diff
@@ -169,10 +169,14 @@ static int get_type(const char *l, const char **t) {
 	return p->type;
 }
 
-static void init_file_tables(void) {
-	static R_TH_LOCAL bool done = false;
+static void init_file_tables(bool *done) {
+	// Initialize
+	if (!done) {
+		done = false;
```
[the remark this answers is not in the document]
Comment (trufae) [reply libr/magic/apprentice.c:175]: please, use pointers properly. or at least check out the compiler warnings before commiting

File: libr/magic/apprentice.c [hunk 9]
```diff
@@ -276,8 +280,9 @@ struct mlist * file_apprentice(RMagic *ms, const char *fn, size_t fn_size, int a
 	char *p, *mfn;
 	int file_err, errs = -1;
 	struct mlist *mlist;
+	bool done = false;
 
-	init_file_tables ();
+	init_file_tables (&done);
```
Comment (trufae) [inline libr/magic/apprentice.c:285]: that will initialize the tables everytime you call file_apprentice. the current implementation initalizes those tables only ONCE per thread and execution, which is not the same.

You should move this initialization into the r_magic_new() and then you can remove the init_file_tables() call inside the file_apprentice.

i have pusehd this change which superseeds this pr https://github.com/radareorg/radare2/pull/20474

**Review by trufae — COMMENTED** [review]
.

**Pull request closed without merging**